File: src/main/java/ru/kainlight/lightcheck/UTILS/Runnables.java
package ru.kainlight.lightcheck.UTILS;

import org.bukkit.OfflinePlayer;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.entity.Player;
import ru.kainlight.lightcheck.API.CheckedPlayer;
import ru.kainlight.lightcheck.common.lightlibrary.CONFIGS.BukkitConfig;
import ru.kainlight.lightcheck.common.lightlibrary.LightPlayer;
import ru.kainlight.lightcheck.Main;

import java.util.*;
import java.util.concurrent.TimeUnit;

public final class Runnables {

    private final Main plugin;

    public Map<OfflinePlayer, String> offlineChecks = new HashMap<>();

    public Map<CheckedPlayer, Integer> messageChatTimer = new HashMap<>();
    public Map<CheckedPlayer, Integer> messageScreenTimer = new HashMap<>();
    public Map<CheckedPlayer, Integer> clockTimerScheduler = new HashMap<>();
    public Map<CheckedPlayer, Integer> bossbarScheduler = new HashMap<>();

    public Runnables(Main plugin) {
        this.plugin = plugin;
    }

    public void start(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        checkedPlayer.startTimer();
        startChatMessageScheduler(checkedPlayer);
        startScreenMessageScheduler(checkedPlayer);
        startBossBarScheduler(checkedPlayer);
    }

    public void stopMessages(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        plugin.getServer().getScheduler().cancelTask(messageChatTimer.get(checkedPlayer));
        plugin.getServer().getScheduler().cancelTask(messageScreenTimer.get(checkedPlayer));
        messageChatTimer.remove(checkedPlayer);
        messageScreenTimer.remove(checkedPlayer);
    }

    public void stopTimer(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        Integer taskID = clockTimerScheduler.remove(checkedPlayer);
        plugin.getServer().getScheduler().cancelTask(taskID);
    }

    public void stopAll(CheckedPlayer checkedPlayer) {
        stopMessages(checkedPlayer);

        if (clockTimerScheduler.get(checkedPlayer) != null) {
            checkedPlayer.stopTimer();
        }
    }

    // TODO: Add an alert for the inspector about the imminent end of the timer or common bossbar
    public void startTimerScheduler(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        if (checkedPlayer.getTimer() != null) {
            clockTimerScheduler.put(checkedPlayer, plugin.getServer().getScheduler().runTaskTimerAsynchronously(plugin, () -> {
                if (checkedPlayer.getPlayer() == null) return;

                if (checkedPlayer.getInspector() == null) {
                    checkedPlayer.disprove();
                    return;
                }

                Long timerValue = checkedPlayer.getTimer();
                if (timerValue >= 1) {
                    checkedPlayer.setTimer(timerValue - 1);
                } else {
                    plugin.getServer().getScheduler().runTask(plugin, () -> {
                        checkedPlayer.approve();
                        List<String> getApproveCommands = plugin.getConfig().getStringList("commands.approve");
                        if (!getApproveCommands.isEmpty()) {
                            getApproveCommands.forEach(commands -> plugin.getServer().dispatchCommand(plugin.getServer().getConsoleSender(), commands.replace("<player>", checkedPlayer.getPlayer().getName())));
                        }
                    });

                    return;
                }

            }, 0, 20L).getTaskId());
        }
    }

    public void startChatMessageScheduler(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        long schedulerTimer = plugin.getConfig().getLong("settings.message-timer") * 20L;
        messageChatTimer.put(checkedPlayer, plugin.getServer().getScheduler().runTaskTimerAsynchronously(plugin, () -> {
            if (checkedPlayer.getPlayer() == null) return;

            this.chatMessage(checkedPlayer);
        }, 0L, schedulerTimer).getTaskId());
    }

    public void startScreenMessageScheduler(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        messageScreenTimer.put(checkedPlayer, plugin.getServer().getScheduler().runTaskTimer(plugin, () -> {
            if (checkedPlayer.getPlayer() == null) return;

            this.screenMessages(checkedPlayer);
        }, 0L, 20L).getTaskId());
    }

    public void startBossBarScheduler(CheckedPlayer checkedPlayer) {
        if (checkedPlayer == null) return;

        Bossbar bossBar = new Bossbar(plugin, checkedPlayer);
        bossbarScheduler.put(checkedPlayer, plugin.getServer().getScheduler().runTaskTimer(plugin, bossBar::show, 0L, 20L).getTaskId());
    }

    private void chatMessage(CheckedPlayer checkedPlayer) {
        Player player = checkedPlayer.getPlayer();
        Player inspector = checkedPlayer.getInspector();

        if (inspector == null || !inspector.isOnline()) {
            checkedPlayer.disprove();
            return;
        }

        String hoverMessage = plugin.getMessageConfig().getConfig().getString("chat.hover");
        if (checkedPlayer.hasTimer()) {
            Long timer = checkedPlayer.getTimer();
            Long secToMin = TimeUnit.SECONDS.toMinutes(timer);
            List<String> with_timer = plugin.getMessageConfig().getConfig().getStringList("chat.with-timer");

            with_timer.forEach(message -> {
                message = message
                        .replace("<inspector>", inspector.getName())
                        .replace("<minutes>", secToMin.toString())
                        .replace("<seconds>", timer.toString());

                LightPlayer.of(player).sendClickableHoverMessage(message, hoverMessage, "/check confirm");
            });
        } else {
            List<String> without_timer = plugin.getMessageConfig().getConfig().getStringList("chat.without-timer");
            without_timer.forEach(message -> {
                message = message.replace("<inspector>", inspector.getName());

                LightPlayer.of(player).sendClickableHoverMessage(message, hoverMessage, "/check confirm");
            });
        }
    }

    private void screenMessages(CheckedPlayer checkedPlayer) {
        Player player = checkedPlayer.getPlayer();
        Player inspector = checkedPlayer.getInspector();

        if (inspector == null || !inspector.isOnline()) {
            checkedPlayer.disprove();
            return;
        }

        Long timer = checkedPlayer.getTimer();
        if (timer != null) {
            boolean titleEnabled = plugin.getConfig().getBoolean("settings.title");
            if (titleEnabled) {
                String titleMessage = plugin.getMessageConfig().getConfig().getString("screen.check-title");
                String subTitleMessage = plugin.getMessageConfig().getConfig().getString("screen.check-subtitle");
                LightPlayer.of(player).sendTitle(titleMessage, subTitleMessage, 1, 15, 2);
            }

            boolean actionbarEnabled = plugin.getConfig().getBoolean("settings.actionbar");
            if (actionbarEnabled) {
                String message = plugin.getMessageConfig().getConfig().getString("screen.actionbar").replace("<seconds>", timer.toString());
                LightPlayer.of(player).sendActionbar(message);
            }
        }

    }

    public void addLog(String username, String text) {
        boolean enabled = plugin.getConfig().getBoolean("settings.logging", false);
        if(!enabled) return;

        text = text.replaceAll("&.", "").replaceAll("#.", "");

        BukkitConfig log = new BukkitConfig(plugin, "logs", username + ".yml", false);
        List<String> list = log.getConfig().getStringList("log");
        list.add(text);

        log.getConfig().set("log", list);
        log.saveConfig();
    }

    public void sendPunishmentCommand(Player player, String alias) {
        final ConsoleCommandSender console = plugin.getServer().getConsoleSender();
        List<String> commands = plugin.getConfig().getStringList("commands." + alias);

        if(!commands.isEmpty()) {
            commands.forEach(command ->
                    plugin.getServer().dispatchCommand(console, command.replace("<player>", player.getName()))
            );
        }
    }


}

File: src/main/java/ru/kainlight/lightcheck/UTILS/Runnables.kt
package ru.kainlight.lightcheck.UTILS

import net.kyori.adventure.text.event.ClickEvent
import org.bukkit.OfflinePlayer
import org.bukkit.command.ConsoleCommandSender
import org.bukkit.entity.Player
import ru.kainlight.lightcheck.API.CheckedPlayer
import ru.kainlight.lightcheck.Main
import ru.kainlight.lightlibrary.LightConfig
import ru.kainlight.lightlibrary.multiActionbar
import ru.kainlight.lightlibrary.multiMessage
import ru.kainlight.lightlibrary.multiTitle
import java.io.File
import java.util.concurrent.TimeUnit

class Runnables(val plugin: Main) {

    val offlineChecks: MutableMap<OfflinePlayer, String> = mutableMapOf()

    val messageChatTimer: MutableMap<CheckedPlayer, Int> = mutableMapOf()
    val messageScreenTimer: MutableMap<CheckedPlayer, Int> = mutableMapOf()
    val clockTimerScheduler: MutableMap<CheckedPlayer, Int> = mutableMapOf()
    val bossbarScheduler: MutableMap<CheckedPlayer, Int> = mutableMapOf()

    fun start(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        checkedPlayer.startTimer()
        startChatMessageScheduler(checkedPlayer)
        startScreenMessageScheduler(checkedPlayer)
        startBossBarScheduler(checkedPlayer)
    }

    fun stopMessages(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        plugin.server.scheduler.cancelTask(messageChatTimer.get(checkedPlayer)!!)
        plugin.server.scheduler.cancelTask(messageScreenTimer.get(checkedPlayer)!!)
        messageChatTimer.remove(checkedPlayer)
        messageScreenTimer.remove(checkedPlayer)
    }

    fun stopTimer(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        val taskID: Int = clockTimerScheduler.remove(checkedPlayer)!!
        plugin.server.scheduler.cancelTask(taskID)
    }

    fun stopAll(checkedPlayer: CheckedPlayer) {
        stopMessages(checkedPlayer)

        if (clockTimerScheduler.get(checkedPlayer) != null) {
            checkedPlayer.stopTimer()
        }
    }

    @Suppress("WARNINGS")
    // TODO: Add an alert for the inspector about the imminent end of the timer or common bossbar
    fun startTimerScheduler(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        if (checkedPlayer.timer != null) {
            val server = plugin.server
            clockTimerScheduler.put(checkedPlayer, server.scheduler.runTaskTimerAsynchronously(plugin, Runnable {
                if (checkedPlayer.player == null) return@Runnable

                if (checkedPlayer.inspector == null) {
                    checkedPlayer.disprove()
                    return@Runnable
                }

                val timerValue: Long = checkedPlayer.timer
                if (timerValue >= 1) {
                    checkedPlayer.timer = timerValue - 1
                } else {
                    server.scheduler.runTask(plugin, Runnable {
                        checkedPlayer.approve()

                        val getApproveCommands: List<String> = plugin.config.getStringList("commands.approve")
                        if (!getApproveCommands.isEmpty()) {
                            getApproveCommands.forEach {
                                server.dispatchCommand(server.consoleSender, it.replace("#player#", checkedPlayer.player.name))
                            }
                        }

                    })
                    return@Runnable
                }
            }, 0L, 20L).taskId)
        }
    }

    fun startChatMessageScheduler(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        val schedulerTimer: Long = plugin.config.getLong("settings.message-timer") * 20L
        messageChatTimer.put(checkedPlayer, plugin.server.scheduler.runTaskTimerAsynchronously(plugin, Runnable {
            if (checkedPlayer.player == null) return@Runnable

            this.chatMessage(checkedPlayer)
        }, 0L, schedulerTimer).taskId)
    }

    fun startScreenMessageScheduler(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        messageScreenTimer.put(checkedPlayer, plugin.server.scheduler.runTaskTimer(plugin, Runnable {
            if (checkedPlayer.player == null) return@Runnable

            this.screenMessages(checkedPlayer)
        }, 0L, 20L).taskId)
    }

    fun startBossBarScheduler(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        val bossBar: Bossbar = Bossbar(plugin, checkedPlayer)
        bossbarScheduler.put(checkedPlayer, plugin.server.scheduler.runTaskTimer(plugin, bossBar::show, 0L, 20L).taskId)
    }

    private fun chatMessage(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        val player: Player = checkedPlayer.player
        val inspector: Player? = checkedPlayer.inspector

        if (inspector == null || !inspector.isOnline) {
            checkedPlayer.disprove()
            return;
        }

        val hoverMessage: String? = plugin.getMessageConfig().getString("chat.hover")
        if (checkedPlayer.hasTimer()) {
            val timer: Long = checkedPlayer.timer
            val secToMin: Long = TimeUnit.SECONDS.toMinutes(timer)
            val with_timer: List<String> = plugin.getMessageConfig().getStringList("chat.with-timer")

            with_timer.forEach {
                var message = it
                message = message
                    .replace("#inspector#", inspector.name)
                    .replace("#minutes#", secToMin.toString())
                    .replace("#seconds#", timer.toString())

                if(hoverMessage != null && !hoverMessage.isBlank()) player.getAudience().multiMessage(message, hoverMessage, ClickEvent.Action.RUN_COMMAND, "/check confirm")
                else player.getAudience().multiMessage(message)
            }
        } else {
            val without_timer: List<String> = plugin.getMessageConfig().getStringList("chat.without-timer")
            without_timer.forEach {
                var message = it
                message = message
                    .replace("#inspector#", inspector.name)

                if(hoverMessage != null && !hoverMessage.isBlank()) player.getAudience().multiMessage(message, hoverMessage, ClickEvent.Action.RUN_COMMAND, "/check confirm")
                else player.getAudience().multiMessage(message)
            }
        }
    }

    private fun screenMessages(checkedPlayer: CheckedPlayer?) {
        if (checkedPlayer == null) return

        val player: Player = checkedPlayer.player
        val inspector: Player? = checkedPlayer.inspector

        if (inspector == null || !inspector.isOnline) {
            checkedPlayer.disprove()
            return;
        }

        val timer: Long? = checkedPlayer.timer
        if (timer != null) {
            val titleEnabled: Boolean = plugin.config.getBoolean("settings.title")
            if (titleEnabled) {
                val titleMessage: String = plugin.getMessageConfig().getString("screen.check-title")!!
                val subTitleMessage: String = plugin.getMessageConfig().getString("screen.check-subtitle")!!
                player.getAudience().multiTitle(titleMessage, subTitleMessage, 1, 15, 2)
            }

            val actionbarEnabled: Boolean = plugin.config.getBoolean("settings.actionbar")
            if (actionbarEnabled) {
                val message: String = plugin.getMessageConfig().getString("screen.actionbar")?.replace("#seconds#", timer.toString())!!
                player.getAudience().multiActionbar(message)
            }
        }
    }

    // $ DEV-7
    fun addLog(username: String, txt: String) {
        val cleanedText = txt
            .replace("""[&][0-9a-fk-or]""".toRegex(), "")
            .replace("""<[^>]+>""".toRegex(), "")
        val enabled: Boolean = plugin.config.getBoolean("settings.logging", false)
        if (!enabled) return

        // !
        val logsFolder = File(plugin.dataFolder, "logs")
        if (!logsFolder.exists()) logsFolder.mkdirs() // Создаем папку logs, если её нет

        val logFile = File(logsFolder, "$username.yml")
        if (!logFile.exists()) {
            logFile.createNewFile() // Создаем файл, если он не существует
        }
        // !

        val logConfig = LightConfig(plugin, "logs", "$username.yml")
        val list: MutableList<String> = logConfig.getConfig().getStringList("log")
        list.add(cleanedText)

        logConfig.getConfig().set("log", list)
        logConfig.saveConfig()
    }

    fun sendPunishmentCommand(player: Player, alias: String) {
        val console: ConsoleCommandSender = plugin.server.consoleSender;
        val commands: List<String> = plugin.config.getStringList("commands.$alias");

        if(commands.isNotEmpty()) {
            commands.forEach {
                plugin.server.dispatchCommand(console, it.replace("#player#", player.name))
            }
        }
    }

}